src/stores/widgets/WidgetLayoutStore.ts

```diff
@@ -0,0 +1,476 @@
+        // unless they have a specific order in mind
+        topWidgets.sort((a, b) => {
+            const layoutA = roomLayout?.widgets?.[a.id];
+            const layoutB = roomLayout?.widgets?.[b.id];
+
+            const userLayoutA = userLayout?.widgets?.[a.id];
+            const userLayoutB = userLayout?.widgets?.[b.id];
+
+            // Jitsi widgets are defaulted to be the leftmost widget whereas other widgets
+            // default to the right side.
+            const defaultA = WidgetType.JITSI.matches(a.type) ? Number.MIN_SAFE_INTEGER : Number.MAX_SAFE_INTEGER;
+            const defaultB = WidgetType.JITSI.matches(b.type) ? Number.MIN_SAFE_INTEGER : Number.MAX_SAFE_INTEGER;
+
+            const orderA = defaultNumber(userLayoutA?.index, defaultNumber(layoutA?.index, defaultA));
+            const orderB = defaultNumber(userLayoutB?.index, defaultNumber(layoutB?.index, defaultB));
+
+            if (orderA === orderB) {
+                // We just need a tiebreak
+                return a.id.localeCompare(b.id);
+            }
+
+            return orderA - orderB;
+        });
+
+        // Determine width distribution and height of the top container now (the only relevant one)
+        const widths: number[] = [];
+        let maxHeight = null; // null == default
+        let doAutobalance = true;
+        for (let i = 0; i < topWidgets.length; i++) {
+            const widget = topWidgets[i];
+            const widgetLayout = roomLayout?.widgets?.[widget.id];
+            const userWidgetLayout = userLayout?.widgets?.[widget.id];
+
+            if (Number.isFinite(userWidgetLayout?.width) || Number.isFinite(widgetLayout?.width)) {
+                const val = userWidgetLayout?.width || widgetLayout?.width;
+                const normalized = clamp(val, MIN_WIDGET_WIDTH_PCT, 100);
+                widths.push(normalized);
+                doAutobalance = false; // a manual width was specified
+            } else {
+                widths.push(100); // we'll figure this out later
+            }
+
+            if (widgetLayout?.height || userWidgetLayout?.height) {
+                const defRoomHeight = defaultNumber(widgetLayout?.height, MIN_WIDGET_HEIGHT_PCT);
+                const h = defaultNumber(userWidgetLayout?.height, defRoomHeight);
+                maxHeight = Math.max(maxHeight, clamp(h, MIN_WIDGET_HEIGHT_PCT, 100));
+            }
+        }
+        if (doAutobalance) {
+            for (let i = 0; i < widths.length; i++) {
+                widths[i] = 100 / widths.length;
+            }
+        }
+
+        // TODO: There is probably a more efficient way to do this.
+        // All we're doing is making sure that our widths sum up to 100 and take
+        // any excess width off all widgets equally to keep the proportions.
+        let toReclaim = sum(...widths) - 100;
+        while (toReclaim > 0 && topWidgets.length > 0) {
```

Comment:
This micro-layout algorithm seems to me like another argument in favour of CSS ratios (via `flex`) plus perhaps some `min-width`, so you can let the browser handle the problem for you.

Reply:
Assuming you continue with the percentages, perhaps there's a two phase approach here you could use instead... Something like:

1. Find all the excess, divide by number of widgets, subtract and clamp that from all of them
2. You might still have a sum above 100 if any hit `MIN_WIDGET_WIDTH_PCT`, so exclude those from future distribution, then repeat step 1

I think that's gets roughly the same answer in less steps, but there could be some flaws I've missed.

Reply:
hmm, yea - this would probably be better than the awful loop I've written